Mathlib/NumberTheory/RamificationInertia/Galois.lean

```diff
@@ -0,0 +1,201 @@
+
+namespace Ideal
+
+open scoped Classical in
+/-- If `L / K` is a Galois extension, it can be seen from the theorem
+  `Ideal.ramificationIdx_eq_of_IsGalois` that all `Ideal.ramificationIdx` over a fixed
+  maximal ideal `p` of `A` are the same, which we define as `Ideal.ramificationIdxIn`. -/
+noncomputable def ramificationIdxIn {A : Type*} [CommRing A] (p : Ideal A)
+    (B : Type*) [CommRing B] [Algebra A B] : ℕ :=
+  if h : ∃ P : Ideal B, P.IsPrime ∧ P.LiesOver p then
+    p.ramificationIdx (algebraMap A B) h.choose else 0
+
+open scoped Classical in
+/-- If `L / K` is a Galois extension, it can be seen from
+  the theorem `Ideal.inertiaDeg_eq_of_IsGalois` that all `Ideal.inertiaDeg` over a fixed
+  maximal ideal `p` of `A` are the same, which we define as `Ideal.inertiaDegIn`. -/
+noncomputable def inertiaDegIn {A : Type*} [CommRing A] (p : Ideal A)
+    (B : Type*) [CommRing B] [Algebra A B] : ℕ :=
+  if h : ∃ P : Ideal B, P.IsPrime ∧ P.LiesOver p then
+    p.inertiaDeg (algebraMap A B) h.choose else 0
+
+section MulAction
+
+variable {A B : Type*} [CommRing A] [CommRing B] [Algebra A B] (p : Ideal A)
+  (K L : Type*) [Field K] [Field L] [Algebra A K] [IsFractionRing A K] [Algebra B L]
+  [Algebra K L] [Algebra A L] [IsScalarTower A B L] [IsScalarTower A K L]
+  [IsIntegralClosure B A L] [FiniteDimensional K L]
+
+instance : MulAction (L ≃ₐ[K] L) (primesOver p B) where
+  smul σ Q := primesOver.mk p (map (galRestrict A K L B σ) Q.1)
+  one_smul Q := by
+    apply Subtype.val_inj.mp
+    show map _ Q.1 = Q.1
+    simpa only [map_one] using map_id Q.1
+  mul_smul σ τ Q := by
+    apply Subtype.val_inj.mp
+    show map _ Q.1 = map _ (map _ Q.1)
+    rw [_root_.map_mul]
+    exact (Q.1.map_map ((galRestrict A K L B) τ).toRingHom ((galRestrict A K L B) σ).toRingHom).symm
+
+theorem coe_smul_primesOver (σ : L ≃ₐ[K] L) (P : primesOver p B):
+    (σ • P).1 = map (galRestrict A K L B σ) P :=
+  rfl
+
+theorem coe_smul_primesOver_mk (σ : L ≃ₐ[K] L) (P : Ideal B) [P.IsPrime] [P.LiesOver p] :
+    (σ • primesOver.mk p P).1 = map (galRestrict A K L B σ) P :=
+  rfl
+
+end MulAction
+
+section RamificationInertia
+
+variable {A B : Type*} [CommRing A] [IsDomain A] [IsIntegrallyClosed A] [CommRing B] [IsDomain B]
+  [IsIntegrallyClosed B] [Algebra A B] [Module.Finite A B] (p : Ideal A) (P Q : Ideal B)
+  [p.IsMaximal] [hPp : P.IsPrime] [hp : P.LiesOver p] [hQp : Q.IsPrime] [Q.LiesOver p]
+  (K L : Type*) [Field K] [Field L] [Algebra A K] [IsFractionRing A K] [Algebra B L]
+  [IsFractionRing B L] [Algebra K L] [Algebra A L] [IsScalarTower A B L] [IsScalarTower A K L]
+
+include p in
+/-- If `p` is a maximal ideal of `A`, `P` and `Q` are prime ideals
+  lying over `p`, then there exists `σ ∈ Aut (B / A)` such that `σ P = Q`. In other words,
+  the Galois group `Gal(L / K)` acts transitively on the set of all prime ideals lying over `p`. -/
+theorem exists_map_eq_of_isGalois [IsGalois K L] : ∃ σ : B ≃ₐ[A] B, map σ P = Q := by
```

Comment:
I have a PR (#17717) that proves transitivity in slightly greater generality (any finite group acting on an extension of rings acts transitively on prime ideals). This might also let you avoid the Galois fraction ring stuff since it seems like this theorem is the only place it gets used?

Reply:
You're right! So I think I can wait for #17717 to be merged and then use `exists_smul_of_under_eq` to replace my theorem `exists_map_eq_of_isGalois`. Have you proven that `[IsInvariant A B (L ≃ₐ[K] L)]` holds if `[IsGalois K L]`?

Reply:
Here you go: https://github.com/leanprover-community/mathlib4/pull/17717/commits/18b9bcce3fe87eeb9dfc53711253502ce26290bd

It's a little messier than I would like, but not too bad.

Reply:
That is exactly what I wanted. Thank you very much!

Reply:
Also, if looks like you've formalized quite a lot of ANT in your neukirch repo. Are you planning on PRing more to mathlib? My PR covers transitivity and surjectivity. But it would be nice to have decomposition and inertia subgroups and the surrounding theory (e.g., the connection with `inertiaDeg` and `ramificationIdx`).

Reply:
Yes. I think I will upstream all the conclusion of [Hilbert's Ramification Theory](https://github.com/jjdishere/neukirch/blob/8fdb75f7ebab8a20eea902058956d7dece648959/AlgebraicNumberTheory/AlgebraicIntegersPart2/HilbertRamificationTheory.lean) in neukirch repo to Mathlib.

Reply:
I've abandoned #17717 in favor of #19926.

Reply:
You can probably remove the dependence on #19141 now that #19926 is merged.

Reply:
> You can probably remove the dependence on #19125 now that #19926 is merged.

Thanks! I think I will open a new PR later, since making changes on this PR might be a bit difficult.